marshal/cluster.go

```diff
@@ -425,16 +433,29 @@ func (c *KafkaCluster) Terminate() {
 	}
 
 	log.Infof("[%s] beginning termination", c.name)
+	marshalers := make([]*Marshaler, 0)
 
+	// This is a bit of a hack, but because marshaler.terminateAndCleanup requires the read lock
+	// on c, we can't terminate the Marshalers in the list while holding the write lock.
+	// Because KafkaCluster::NewMarshaler will return an error if the marshalers slice is nil,
+	// we know there cannot be new Marshalers created which aren't included in the local slice
+	// we create here.
+	//
+	// There is probably some alternative where the quit variable is protected by the mutex instead
+	// of being atomic, but this seems somewhat cleaner in case some future refactoring eliminates
+	// the marshalers slice entirely this hack will go away automatically.
 	c.lock.Lock()
-	defer c.lock.Unlock()
-
 	// Terminate all Marshalers which will in turn terminate all Consumers and
 	// let everybody know we're all done.
 	for _, marshaler := range c.marshalers {
-		marshaler.terminateAndCleanup(false)
+		marshalers = append(marshalers, marshaler)
```

Comment:
Couldn't you just,

```
c.lock.Lock()
marshalers := c.marshalers
c.marshalers = nil
c.lock.Unlock()
```

i.e., is there any reason to do the append dance onto a new slice? Are you worried about someone else having a copy of the slice while we're mutating it? (Shouldn't be possible I think?)